File: rust/src/dtn/store.rs
// SPDX-License-Identifier: MIT

use async_trait::async_trait;
use std::collections::HashMap;
use std::sync::Mutex;
use uuid::Uuid;

use crate::models::{BundleStatus, CustodyRecord, DtnBundle};

/// Persistent backing store for DTN bundles + custody records.
#[async_trait]
pub trait BundleStore: Send + Sync {
    async fn get(&self, bundle_id: &Uuid) -> Option<DtnBundle>;
    async fn get_active(&self) -> Vec<DtnBundle>;
    async fn save(&self, bundle: DtnBundle);
    async fn remove(&self, bundle_id: &Uuid);
    async fn get_active_count(&self) -> usize;
    async fn save_custody(&self, record: CustodyRecord);
    async fn get_custody_records(&self, bundle_id: &Uuid) -> Vec<CustodyRecord>;
    async fn expire_stale(&self) -> usize;
}

/// Process-local DTN store. Suitable for tests.
pub struct InMemoryBundleStore {
    bundles: Mutex<HashMap<Uuid, DtnBundle>>,
    custody: Mutex<HashMap<Uuid, CustodyRecord>>,
}

impl Default for InMemoryBundleStore {
    fn default() -> Self {
        Self::new()
    }
}

impl InMemoryBundleStore {
    pub fn new() -> Self {
        Self {
            bundles: Mutex::new(HashMap::new()),
            custody: Mutex::new(HashMap::new()),
        }
    }
}

#[async_trait]
impl BundleStore for InMemoryBundleStore {
    async fn get(&self, bundle_id: &Uuid) -> Option<DtnBundle> {
        self.bundles.lock().unwrap().get(bundle_id).cloned()
    }

    async fn get_active(&self) -> Vec<DtnBundle> {
        self.bundles
            .lock()
            .unwrap()
            .values()
            .filter(|b| {
                !b.is_expired()
                    && (b.status == BundleStatus::Pending || b.status == BundleStatus::InCustody)
            })
            .cloned()
            .collect()
    }

    async fn save(&self, bundle: DtnBundle) {
        self.bundles.lock().unwrap().insert(bundle.id, bundle);
    }

    async fn remove(&self, bundle_id: &Uuid) {
        self.bundles.lock().unwrap().remove(bundle_id);
    }

    async fn get_active_count(&self) -> usize {
        self.bundles
            .lock()
            .unwrap()
            .values()
            .filter(|b| {
                !b.is_expired()
                    && (b.status == BundleStatus::Pending || b.status == BundleStatus::InCustody)
            })
            .count()
    }

    async fn save_custody(&self, record: CustodyRecord) {
        self.custody.lock().unwrap().insert(record.id, record);
    }

    async fn get_custody_records(&self, bundle_id: &Uuid) -> Vec<CustodyRecord> {
        self.custody
            .lock()
            .unwrap()
            .values()
            .filter(|r| r.bundle_id == *bundle_id)
            .cloned()
            .collect()
    }

    async fn expire_stale(&self) -> usize {
        let mut bundles = self.bundles.lock().unwrap();
        let mut expired = 0;
        for b in bundles.values_mut() {
            if b.is_expired() && b.status != BundleStatus::Expired {
                b.status = BundleStatus::Expired;
                expired += 1;
            }
        }
        expired
    }
}

File: go/dtn/store.go
// SPDX-License-Identifier: MIT

// Package dtn implements delay-tolerant networking on top of the Aether mesh.
// Bundles are JSON-encoded into MeshPacket payloads, replicated to nearby peers
// via an epidemic strategy (default: geohash-proximity), and delivered when the
// recipient appears or accepts custody from a closer carrier.
package dtn

import (
	"context"
	"sync"

	"github.com/bhengubv/aether-protocol/go/models"
)

// BundleStore is the persistent backing store for DTN bundles + custody records.
type BundleStore interface {
	Get(ctx context.Context, bundleID string) (*models.DtnBundle, error)
	GetActive(ctx context.Context) ([]models.DtnBundle, error)
	Save(ctx context.Context, bundle *models.DtnBundle) error
	Remove(ctx context.Context, bundleID string) error
	GetActiveCount(ctx context.Context) (int, error)
	SaveCustody(ctx context.Context, record *models.CustodyRecord) error
	GetCustodyRecords(ctx context.Context, bundleID string) ([]models.CustodyRecord, error)
	ExpireStale(ctx context.Context) (int, error)
}

// InMemoryBundleStore is a process-local store. Loses data on restart.
type InMemoryBundleStore struct {
	mu       sync.RWMutex
	bundles  map[string]*models.DtnBundle
	custody  map[string]*models.CustodyRecord
}

// NewInMemoryBundleStore creates a fresh empty store.
func NewInMemoryBundleStore() *InMemoryBundleStore {
	return &InMemoryBundleStore{
		bundles: make(map[string]*models.DtnBundle),
		custody: make(map[string]*models.CustodyRecord),
	}
}

func (s *InMemoryBundleStore) Get(ctx context.Context, bundleID string) (*models.DtnBundle, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	b, ok := s.bundles[bundleID]
	if !ok {
		return nil, nil
	}
	return b, nil
}

func (s *InMemoryBundleStore) GetActive(ctx context.Context) ([]models.DtnBundle, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	out := make([]models.DtnBundle, 0, len(s.bundles))
	for _, b := range s.bundles {
		if b.IsExpired() {
			continue
		}
		if b.Status == models.DtnStatusPending || b.Status == models.DtnStatusInCustody {
			out = append(out, *b)
		}
	}
	return out, nil
}

func (s *InMemoryBundleStore) Save(ctx context.Context, bundle *models.DtnBundle) error {
	if bundle == nil {
		return nil
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	s.bundles[bundle.ID] = bundle
	return nil
}

func (s *InMemoryBundleStore) Remove(ctx context.Context, bundleID string) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.bundles, bundleID)
	return nil
}

func (s *InMemoryBundleStore) GetActiveCount(ctx context.Context) (int, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	count := 0
	for _, b := range s.bundles {
		if b.IsExpired() {
			continue
		}
		if b.Status == models.DtnStatusPending || b.Status == models.DtnStatusInCustody {
			count++
		}
	}
	return count, nil
}

func (s *InMemoryBundleStore) SaveCustody(ctx context.Context, record *models.CustodyRecord) error {
	if record == nil {
		return nil
	}
	s.mu.Lock()
	defer s.mu.Unlock()
	s.custody[record.ID] = record
	return nil
}

func (s *InMemoryBundleStore) GetCustodyRecords(ctx context.Context, bundleID string) ([]models.CustodyRecord, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	out := make([]models.CustodyRecord, 0)
	for _, r := range s.custody {
		if r.BundleID == bundleID {
			out = append(out, *r)
		}
	}
	return out, nil
}

func (s *InMemoryBundleStore) ExpireStale(ctx context.Context) (int, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	expired := 0
	for _, b := range s.bundles {
		if b.IsExpired() && b.Status != models.DtnStatusExpired {
			b.Status = models.DtnStatusExpired
			expired++
		}
	}
	return expired, nil
}
